lib/src/loader.rs

```diff
@@ -105,19 +106,51 @@ fn load_elf(fd: &mut File, name: String) -> Result<(Project,Machine)> {
 
     prog.call_graph.add_vertex(CallTarget::Todo(Rvalue::new_u64(entry as u64),Some(name),Uuid::new_v4()));
 
-    let dynstrtab = binary.dynstrtab;
-    for sym in binary.dynsyms {
-        let name = dynstrtab[sym.st_name() as usize].to_string();
-        let addr = sym.st_value();
-        debug!("{} @ 0x{:x}: {:?}", name, addr, sym);
+    let add_sym = |prog: &mut Program, sym: &elf::Sym, strtab: &goblin::strtab::Strtab| {
+        let name = strtab[sym.st_name].to_string();
+        let addr = sym.st_value;
+        debug!("Symbol: {} @ 0x{:x}: {:?}", name, addr, sym);
         if sym.is_function() {
             if sym.is_import() {
                 prog.call_graph.add_vertex(CallTarget::Symbolic(name,Uuid::new_v4()));
             } else {
                 prog.call_graph.add_vertex(CallTarget::Todo(Rvalue::new_u64(addr),Some(name),Uuid::new_v4()));
             }
         }
+    };
+
+    let resolve_import_address = |proj: &mut Project, relocs: &[elf::Reloc], name: &str| {
+        for reloc in relocs {
+            let pltsym = &binary.dynsyms[reloc.r_sym];
+            let pltname = &binary.dynstrtab[pltsym.st_name];
+            if pltname == name {
+                debug!("Import match {}: {:#x} {:?}", name, reloc.r_offset, pltsym);
+                proj.imports.insert(reloc.r_offset as u64, name.to_string());
+                return true;
+            }
+        }
+        false
+    };
+
+    // add dynamic symbol information (non-strippable)
+    for sym in &binary.dynsyms {
+        add_sym(&mut prog, sym, &binary.dynstrtab);
+        if sym.is_function () {
+            let name = &binary.dynstrtab[sym.st_name];
+            if !resolve_import_address(&mut proj, &binary.pltrelocs, name) {
+                if !resolve_import_address(&mut proj, &binary.dynrelas, name) {
+                    resolve_import_address(&mut proj, &binary.dynrels,  name);
```

Comment:
I suppose I could have, but its specific to ELF imports (e.g., the `&[Reloc]` argument), and can't be used with mach-o, or PE imports, and so I felt like hiding it in the ELF loader implementation details. Also doing so would expose goblin internals in the `Project` api, which doesn't seem necessary (i.e., I think its probably best for binary loading details to all be private functions called by the public and generic loader)

Speaking of mach-o, I've got a preliminary mach-o loader done on a local branch, and looks like imports are very similar (with a jump stub); I haven't thought about or looked at PE yet, but I suspect what we have here will work as a base.

So question is how we're going to use this import table to actually collapse the `callq`s into the names we have here (which is another issue, but wondering if we have anything in the works (I'm secretly hoping @flanfly has it all done already :laughing: ))?